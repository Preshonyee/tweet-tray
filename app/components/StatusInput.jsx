import React, { Component, } from 'react';
import { connect, } from 'react-redux';
import PropTypes from 'prop-types';
import TwitterText from 'twitter-text';
import Styled from 'styled-components';
import Theme from 'styled-theming';

import ConnectUtilities from '../containers/ConnectUtilities';

import * as constants from '../constants';

import { updateWeightedStatus, } from '../actions';

const StatusInputStyle = Styled.div`
    width: auto;
    user-select: auto;
    margin-top: 5px;
    margin-left: 66px;
    margin-bottom: ${constants.SPACING}px;
    background-color: transparent;

    .TextArea {
        line-height: 26px;
        outline: 0;
        border: 0;
        margin: 0;
        padding: 0;
        width: 100%;
        height: auto;
        resize: none;
        overflow-y: auto;
        font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif !important;
        font-size: ${constants.LARGE_FONT_SIZE}px;
        font-weight: normal;
        color:  ${Theme('mode', { day: constants.BLACK, night: constants.WHITE, })};
        background-color: transparent;

        &::placeholder {
          color: ${constants.GREY};
        }
    }
`;

class StatusInput extends Component {
  static propTypes = {
    placeholder: PropTypes.string,
    weightedStatusText: PropTypes.string,
    updateWeightedStatus: PropTypes.func.isRequired,
  }

  static defaultProps = {
    placeholder: '',
    weightedStatusText: '',
  }

  constructor(props) {
    super(props);

    this._adjustTextarea = this._adjustTextarea.bind(this);
    this._focusTextArea = this._focusTextArea.bind(this);
    this._onTextAreaUpdate = this._onTextAreaUpdate.bind(this);
  }

  componentDidMount() {
    this._adjustTextarea({});
    this._focusTextArea({});
  }

  _adjustTextarea({ target = this.el, }) {
    const textAreaRef = target;
    if (textAreaRef !== undefined && textAreaRef !== null) {
      textAreaRef.style.height = 0;
      textAreaRef.style.height = `${target.scrollHeight}px`;
    }
  }

  _focusTextArea({ target = this.el, }) {
    const textAreaRef = target;

    if (textAreaRef !== undefined && textAreaRef !== null) {
      textAreaRef.focus();
    }
  }

  _onTextAreaUpdate(e) {
    const { updateWeightedStatus, } = this.props;
    this._adjustTextarea(e);

    const textValue = e.target.value;
    const parsedProperties = TwitterText.parseTweet(textValue);

    const status = {
      text: textValue,
      weightedLength: parsedProperties.weightedLength,
      permillage: parsedProperties.permillage,
      valid: parsedProperties.valid,
      displayRangeStart: parsedProperties.displayRangeStart,
      displayRangeEnd: parsedProperties.displayRangeEnd,
      validDisplayRangeStart: parsedProperties.validDisplayRangeStart,
      validDisplayRangeEnd: parsedProperties.validDisplayRangeEnd,
    };

    if (status.text.length !== 0) {
      updateWeightedStatus(status);
    } else {
      updateWeightedStatus(null);
    }
  }

  render() {
    const { placeholder, weightedStatusText, } = this.props;
    return (
      <StatusInputStyle>
        <textarea
          ref={(x) => { this.el = x; }}
          className="textArea"
          autoCapitalize="sentences"
          placeholder={placeholder}
          rows={1}
          style={{
            minHeight: 58,
          }}
          onInput={this._onTextAreaUpdate}
          onKeyUp={this._onTextAreaUpdate}
          onChange={this._onTextAreaUpdate}
          defaultValue={weightedStatusText}
        />
      </StatusInputStyle>
    );
  }
}

export default ConnectUtilities(StatusInput);

