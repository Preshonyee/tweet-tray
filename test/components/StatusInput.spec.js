import React from 'react';
import Enzyme, { shallow, } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import StatusInput from '../../app/components/StatusInput';
import * as testData from '../testData';

Enzyme.configure({adapter: new Adapter()});

function setup() {
  const component = shallow(<StatusInput
    placeholder={'What\'s happening?'}
    weightedStatusText={testData.weightedStatus.text}
    updateWeightedStatus={(status) => { return status; }}
  />);
  return {
    component,
    textarea: component.find('textarea'),
  };
}

describe('StatusInput component', () => {
  it('should display textarea value', () => {
    const { textarea, } = setup();
    expect(textarea.prop('defaultValue')).toMatchSnapshot();
  });
});
