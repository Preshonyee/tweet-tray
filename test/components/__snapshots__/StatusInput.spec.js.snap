// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusInput component should display textarea value 1`] = `"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam non nisl facilisis, imperdiet ipsum sed, pretium metus. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos."`;
